test/expr/closure/closures.swift

```diff
@@ -733,6 +739,36 @@ public class TestImplicitCaptureOfExplicitCaptureOfSelfInEscapingClosure {
     }
 }
 
+public class TestImplicitSelfForWeakSelfCapture {
+  static var staticOptional: TestImplicitSelfForWeakSelfCapture? = .init()
+  func method() { }
+  
+  private init() {
+    doVoidStuff { [weak self] in
+      guard let self = self else { return }
+      method()
```

Comment:
Could you please add a test where `method()` is used before `guard let` as well just to make sure it's diagnosed?